assembly/index.ts

```diff
@@ -0,0 +1,195 @@
+
+const hexAlphabet = '0123456789abcdef';
+
+class Decoded {
+    data: RLPData;
+    remainder: Uint8Array;
+
+    constructor(data: RLPData, remainder: Uint8Array) {
+        this.data = data;
+        this.remainder = remainder;
+    }
+}
+
+/**
+ * Parse integers. Check if there is no leading zeros
+ * Note that this is NOT safe in assemblyscript due to
+ * the lack of error handling.
+ * @param v The value to parse
+ * @param base The base to parse the integer into
+ */
+function safeParseInt(v: string, base: u32): u32 {
+    if (v.slice(0, 2) == '00') {
+        throw new Error('invalid RLP: extra zeros');
+    }
+    return parseI32(v, base) as u32;
+}
+
+/** Transform an integer into its hexadecimal value */
+function intToHex(integer: u32): string {
+    let res = new Array<string>();
+    do {
+        let t = integer / 16;
+        let r = integer % 16;
+        integer = t;
+        res.push(hexAlphabet[r]);
+    } while (integer);
+    let hex = res.reverse().join('');
+    return hex.length % 2 ? "0" + hex : hex;
+}
+
+function bytesToHex(bytes: Uint8Array): string {
+    let arr = new Array<string>();
+    for (let i = 0; i < bytes.length; i++) {
+        arr.push(intToHex(bytes[i]));
+    }
+    return arr.join('');
+}
+
+function hexToBytes(hex: string): Uint8Array {
+    if (!hex.length) {
+        return null;
+    }
+    assert(hex.length % 2 == 0);
+    let byte_length = hex.length / 2;
+    let res = new Uint8Array(byte_length);
+    for (let i = 0; i < byte_length; i++) {
+        res[i] = parseI32(hex.substr(i*2, 2), 16) as u8;
+    }
+    return res;
+}
+
+function concatUint8Array(arr1: Uint8Array, arr2: Uint8Array): Uint8Array {
+    let len = arr1.length + arr2.length;
+    let res = new Uint8Array(len);
+    for (let i = 0; i < len; i++) {
+        if (i < arr1.length) {
+            res[i] = arr1[i];
+        } else {
+            res[i] = arr2[i-arr1.length];
+        }
+    }
+    return res;
+}
+
+function concatUint8Arrays(arrays: Array<Uint8Array>): Uint8Array {
+    let len = arrays.reduce<u32>(((acc, x) => acc + x.length), 0);
+    let res = new Uint8Array(len);
+    let counter = 0;
+    for (let i = 0; i < arrays.length; i++) {
+        for (let j = 0; j < arrays[i].length; j++) {
+            res[counter++] = arrays[i][j];
+        }
+    }
+    return res;
+}
+
+/**
+ * RLP Encoding based on: https://github.com/ethereum/wiki/wiki/%5BEnglish%5D-RLP
+ * This function takes in an argument of type Input and returns the rlp encoding of it.
+ * @param input: a Uint8Array or an array of inputs.
+ * @returns returns rlp encoded byte array.
+ **/
+export function encode(input: RLPData): Uint8Array {
+    if (input.children) {
+        let output = new Array<Uint8Array>();
+        for (let i = 0; i < input.children.length; i++) {
+            output.push(encode(input.children[i]));
+        }
+        let buf = concatUint8Arrays(output);
+        return concatUint8Array(encodeLength(buf.length, 192), buf);
```

Comment:
Upon closer inspection, I don't think writing at some offset really solves the problem because in the case of a list of items, computing the offset requires one to go through the list, computing the encoding of each item, which itself already produces those `Uint8Array` we try to avoid. Due to the recursive nature of `encode`, I don't see a way around that.

Reply:
@bowenwang1996 hmm, you can just overwrite length after recursing, see e.g. https://github.com/nearprotocol/assemblyscript-bson/blob/master/assembly/encoder.ts#L78